controller/tests/controller_server/csi_controller_server_test.py

```diff
@@ -555,148 +563,208 @@ def test_create_volume_with_zero_size(self, storage_agent, array_type):
+        source_id = "wwn1"
+        target_volume_id = "wwn2"
+        self.request.volume_content_source = self._get_source_snapshot(source_id)
+        self.mediator.create_volume = Mock()
+        self.mediator.create_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name,
+                                                                                           target_volume_id, "a9k")
+        self.mediator.get_object_by_id = Mock()
+        if not copy_exception:
+            self.mediator.get_object_by_id.side_effect = [get_exception]
+            self.storage_agent.get_mediator.return_value.__exit__.side_effect = [get_exception]
+        else:
+            self.mediator.get_object_by_id.return_value = utils.get_mock_mediator_response_snapshot(1000, snapshot_name,
+                                                                                                    target_volume_id,
+                                                                                                    volume_name, "a9k")
+            self.mediator.copy_to_existing_volume_from_source = Mock()
+            self.mediator.copy_to_existing_volume_from_source.side_effect = [copy_exception]
+
+            self.storage_agent.get_mediator.return_value.__exit__.side_effect = [copy_exception]
         self.mediator.delete_volume = Mock()
         array_type.return_value = "a9k"
         self.servicer.CreateVolume(self.request, self.context)
-        if rollback_called:
-            self.mediator.delete_volume.assert_called_with(vol_id)
+        self.mediator.delete_volume.assert_called_with(target_volume_id)
         self.assertEqual(self.context.code, return_code)
 
+    @patch("controller.controller_server.csi_controller_server.detect_array_type")
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_clone_volume_success(self, storage_agent, array_type):
+        storage_agent.return_value = self.storage_agent
+        volume_id = "wwn1"
+        volume_capacity_bytes = 100
+        array_type.return_value = "a9k"
+        self.request.volume_content_source = self._get_source_volume(volume_id)
+        self.mediator.create_volume = Mock()
+        self.mediator.create_volume.return_value = utils.get_mock_mediator_response_volume(10, volume_name, "wwn2",
+                                                                                           "a9k")
+        self.mediator.get_object_by_id = Mock()
+        self.mediator.get_object_by_id.return_value = utils.get_mock_mediator_response_volume(volume_capacity_bytes,
+                                                                                              clone_volume_name,
+                                                                                              volume_id,
+                                                                                              "a9k")
+        self.mediator.copy_to_existing_volume_from_source = Mock()

```

Comment:
could this be moved to a common "prepare mocks" function?

Reply:
done

Reply:
can't it also contain assignments to `self.request.volume_content_source`, `self.mediator.create_volume.return_value`, etc...?

Reply:
done (everything else change between clone and snapshot)

Reply:
I thought maybe add some more (as default values) and overwrite them where necessary, but ok